Implement label API

grouville opened the pull request:
Closes https://github.com/dagger/dagger/issues/4245
Closes https://github.com/dagger/dagger/issues/3887

Allows user to have the complete control over the labels of a given container:
- Add a label (`withLabel` API)
- Remove a label (`withoutLabel` API)
- Print the value of a single label (`label` API)
- Print all the labels of a container (`labels` API)

Adds corresponding graphQL tests, SDK codegen and graphQL docs.

The terminology is consistent with other APIs (cf. envVariable), and will help solving inherent docker issue https://github.com/moby/moby/issues/3465

aluzzardi commented on core/integration/container_test.go:
```diff
@@ -932,6 +932,115 @@ func TestContainerEnvVariablesReplace(t *testing.T) {
 	require.Contains(t, res.Container.From.WithEnvVariable.WithExec.Stdout, "GOPATH=/gone\n")
 }
 
+func TestContainerWithLabel(t *testing.T) {
```
nit: Feel free to use the Go SDK in these tests. The reason the other tests are not using the SDK it's just because they were written before the SDK existed. For new tests though, the SDK is preferred

grouville replied:
Done, great idea. As we found out about this bug: https://github.com/dagger/dagger/issues/4398

aluzzardi reviewed (approved):
LGTM

/cc @vito for final review

vito commented on core/schema/container.graphqls:
```diff
@@ -88,6 +88,18 @@ type Container {
+  "Retrieves this container plus the given label."
+  withLabel(name: String!, value: String!): Container!
+
+  "Retrieves this container minus the given environment variable."
```
missed a spot!

```suggestion
  "Retrieves this container minus the given label."
```

grouville replied:
thanks 🙏

vito commented on core/schema/container.go:
```diff
@@ -406,6 +451,32 @@ func (s *containerSchema) mounts(ctx *router.Context, parent *core.Container, _
+
+func (s *containerSchema) withoutLabel(ctx *router.Context, parent *core.Container, args containerWithoutLabelArgs) (*core.Container, error) {
+	return parent.UpdateImageConfig(ctx, func(cfg specs.ImageConfig) specs.ImageConfig {
+		delete(cfg.Labels, args.Name)
```
Will this panic with nil `cfg.Labels`?

vito commented on core/schema/container.go:
```diff
@@ -344,6 +348,47 @@ func (s *containerSchema) envVariable(ctx *router.Context, parent *core.Containe
+		return nil, err
+	}
+
+	if val, ok := cfg.Labels[args.Name]; ok {
```
Possible panic here too?

vito commented on core/schema/container.go:
```diff
@@ -344,6 +348,47 @@ func (s *containerSchema) envVariable(ctx *router.Context, parent *core.Containe
+	}
+
+	labels := make([]Label, 0, len(cfg.Labels))
+	for name, value := range cfg.Labels {
```
And maybe here?

vito reviewed (commented):
Looks great! I have some mild paranoia around nil panics, and annotated each spot I think it could possibly maybe happen. Could we have some extra tests to cover that? Basically the tests you have already, but without the `from`.

grouville commented:
> Looks great! I have some mild paranoia around nil panics, and annotated each spot I think it could possibly maybe happen. Could we have some extra tests to cover that? Basically the tests you have already, but without the `from`.

- Updated all tests using Go SDK, apart from the `labels` endpoint, which remains as a GraphQL test, as it faces a tangent issue: https://github.com/dagger/dagger/issues/4398.
- Regarding the nil: those tests were already manually tested, but adding them to the test suite is a great idea 🙏
- re-generated the API, as there was a typo
- rebased on top of main 🙏

TomChv commented on core/schema/container.graphqls:
```diff
@@ -88,6 +88,18 @@ type Container {
   "Retrieves this container plus the given environment variable."
   withEnvVariable(name: String!, value: String!): Container!
 
+  "Retrieves the list of labels passed to container."
```
Nit: for consistency
```suggestion
  "Retrieves the list of labels passed to this container."
```

grouville replied:
I checked on all the codebase, there is no mention of `to this` anywhere. I followed the consistency of envVariables

TomChv reviewed (approved):
LGTM! I just left a nit comment on the GQL schema :rocket:

grouville commented:
@vito, rebased and implemented the changes requested by Andrea. I let the `labels` tests as graphQL query, for the issue mentioned above. Shall be on mergeable state, just wanted a last check from you 😇 🙏

vito reviewed (approved):
lgtm!
